src/Microsoft.Diagnostics.Runtime/src/Utilities/PEImage/PEImage.cs

```diff
@@ -315,71 +286,95 @@ private string ReadString(int offset, int len)
 
             SeekTo(offset);
 
-            byte[] buffer = GetBuffer(len);
-            if (Stream.Read(buffer, 0, len) != len)
-                return null;
-
-            for (int i = 0; i < len; i++)
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(len);
+            try
             {
-                if (buffer[i] == 0)
+                if (Stream.Read(buffer, 0, len) != len)
+                    return null;
+
+                for (int i = 0; i < len; i++)
                 {
-                    len = i;
-                    break;
+                    if (buffer[i] == 0)
+                    {
+                        len = i;
+                        break;
+                    }
                 }
+
+                return Encoding.ASCII.GetString(buffer, 0, len);
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
             }
-            
-            return Encoding.ASCII.GetString(buffer, 0, len);
         }
 
-        private T? TryRead<T>() where T : struct
-        {
-            return TryRead<T>(_offset);
-        }
+        private bool TryRead<T>(out T result) where T : unmanaged => TryRead(_offset, out result);
 
-        private T? TryRead<T>(int offset) where T : struct
+        private bool TryRead<T>(int offset, out T t) where T : unmanaged
         {
-            int size = Marshal.SizeOf(typeof(T));
-            byte[] buffer = GetBuffer(size);
+            t = default;
+            int size = Unsafe.SizeOf<T>();
 
-            SeekTo(offset);
-            int result = Stream.Read(buffer, 0, size);
-            _offset = offset + result;
+            // .Net Core only.  This isn't compiled into Desktop CLR because Stream.Read(Span<byte>) doesn't
+            // exist, and so our stackalloc + copy will be more inefficent than just renting our own byte array.
+#if CORE_ONLY
+            if (size < Configuration.MaxStackAlloc)
+            {
+                byte* ptr = stackalloc byte[size];
```

Comment:
If you now or in the future disable zero'ing of locals (today using the ILLink feature for it, or in the near future using the Roslyn feature for it), then it'll be more efficient to instead do:
```C#
stackalloc byte[Configuration.MaxStackAlloc]
```
It's a bit cheaper for the JIT to stackalloc a const size than it is a dynamic size.

Reply:
How about:
```C#
Span<byte> span = stackalloc byte[size];
```
?

Then below instead of:
```C#
Unsafe.Copy(ref t, ptr);
```
you can do:
```C#
t = MemoryMarshal.Read<T>(span);
```
?

Reply:
There is a slight difference between these two methods.

How about:

```
t = Unsafe.As<byte, T>(ref span[0]);
```

?

Reply:
There is a slight difference between these two methods.

How about:

```csharp
t = Unsafe.As<byte, T>(ref span[0]);
```

?

Reply:
> How about

Why? MemoryMarshal.Read uses Unsafe.ReadUnaligned on coreclr (and just Unsafe.As, otherwise)... what's your concern with that?

Reply:
`Unsafe.ReadUnaligned` has an additional `unaligned` instruction.

https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il#L85

https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il#L313

Reply:
`Unsafe.ReadUnaligned` contains an additional `unaligned` instruction.

https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il#L85

https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il#L313

Reply:
Right. Why are you suggesting the one that doesn't have it?

Reply:
I assumed that `localloc` shall allocate aligned addresses. Not sure if it still holds for the `byte` target type.